pdns/dnsdist-cache.cc

```diff
@@ -276,6 +276,24 @@ void DNSDistPacketCache::expungeByName(const DNSName& name, uint16_t qtype)
   }
 }
 
+void DNSDistPacketCache::expungeBySuffix(const DNSName& name, uint16_t qtype)
+{
+    WriteLock w(&d_lock);
+
+    for (auto it = d_map.begin(); it != d_map.end(); ) {
+        const CacheValue& value = it->second;
+        uint16_t cqtype = 0;
+        uint16_t cqclass = 0;
+        DNSName cqname(value.value.c_str(), value.len, sizeof(dnsheader), false, &cqtype, &cqclass, nullptr);
```

Comment:
The `qname`, `qtype` and `qclass` are stored in the cache entry, we don't need to parse them again. I just noticed `DNSDistPacketCache::expungeByName()` does the parsing too, which is awful, we should fix that!
Thinking about it a bit more, we should probably just add a boolean parameter to `DNSDistPacketCache::expungeByName()` and do an equal comparison or an `isPartOf()` one depending on the boolean value.